src/core/geometry/qgsgeos.cpp

```diff
@@ -548,7 +586,21 @@ double QgsGeos::distance( const QgsAbstractGeometry *geom, QString *errorMsg ) c
     return distance;
   }
 
-  geos::unique_ptr otherGeosGeom( asGeos( geom, mPrecision ) );
+  geos::unique_ptr otherGeosGeom;
+
+  // GEOSPreparedDistance_r is not able to properly compute the distance if one
+  // of the geometries if a vertical line (LineString Z((X Y Z1, X Y Z2, ..., X Y Zn))).
+  // In that case, replace `geom` by a single point.
+  // However, GEOSDistance_r works.
+  if ( mGeosPrepared && isZVerticalLine( geom->simplifiedTypeRef() ) )
+  {
+    otherGeosGeom = asGeos( geom->vertexAt( QgsVertexId( 0, 0, 0 ) ).clone(), mPrecision );
```

Comment:
asGeos doesn't take ownership, so this leaks the cloned point

Reply:
Fixed.